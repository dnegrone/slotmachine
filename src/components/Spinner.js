import React, { Component } from 'react';

class Spinner extends Component {  
    constructor(props){
      super(props);
      this.forceUpdateHandler = this.forceUpdateHandler.bind(this);
    };
  
    state = {
      position: 0,
      lastPosition: null
    }
    forceUpdateHandler() {
      this.reset();
    }; 
    
    reset() {
      // if timer exists, reset timer
      if (this.timer) { 
        clearInterval(this.timer); 
      }  
      
      // setting the start position
      this.start = this.setStartPosition();
      
      // reseting the position and the timer
      this.setState({
        position: this.start,
        timeRemaining: this.props.timer        
      });
      
      // initializing timer
      this.timer = setInterval(() => {
        this.tick()
      }, 100);     
    }
    // getting the height of the icon without restarting the component
    static iconHeight = 188;
    
    // setting the speed based in this current position
    multiplier = Math.floor(Math.random()*(4-1)+1);  
    start = this.setStartPosition();
    speed = Spinner.iconHeight * this.multiplier;    
    
    // positioning the height of the sprite 
    setStartPosition() {
      return ((Math.floor((Math.random()*9))) * Spinner.iconHeight)*-1;
    }
  
    // moving the background ad
    moveBackground() {
      this.setState({ 
        position: this.state.position - this.speed,
        timeRemaining: this.state.timeRemaining - 100
      })
    }
  
    getSymbolFromPosition() {
      let { position } = this.state;
      const totalSymbols = 9;
      const maxPosition = (Spinner.iconHeight * (totalSymbols-1)*-1);
      let moved = (this.props.timer/100) * this.multiplier
      let startPosition = this.start;
      let currentPosition = startPosition;    
  
      for (let i = 0; i < moved; i++) {              
        currentPosition -= Spinner.iconHeight;
  
        if (currentPosition < maxPosition) {
          currentPosition = 0;
        }      
      }
      
      // This enable the button <RepeatButton> to be triggered and validating the current position and message
      this.props.onFinish(currentPosition);
      // return this.props.onFinish(currentPosition);
    }
  
    tick() {      
      if (this.state.timeRemaining <= 0) {
        clearInterval(this.timer);        
        this.getSymbolFromPosition();    
  
      } else {
        this.moveBackground();
      }      
    }
  
    componentDidMount() {
      clearInterval(this.timer);
  
      this.setState({
        position: this.start,
        timeRemaining: this.props.timer
      });
  
      this.timer = setInterval(() => {
        this.tick()
      }, 100);
    }
  
    render() {
      let { position, current } = this.state;
  
      return (            
        <div 
          style={{backgroundPosition: '0px ' + position + 'px'}}
          className={`icons`}          
        />
      )
    }
}

export default Spinner;