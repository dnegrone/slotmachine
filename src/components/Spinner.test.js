import React from 'react';
import { shallow } from 'enzyme';
import Spinner from './Spinner';

describe('Spinner Tests', () => {
    let wrapper;
    const mockFunctions = jest.fn();
    beforeEach(() => {
        const mockProps = {};
        wrapper = shallow(<Spinner {...mockProps} />);
    });
    
    test('renders MainPage without crashing', () => {
        // console.log(wrapper.debug())
        expect(wrapper.debug()).toMatchSnapshot();
    })

    test('expect to see all current states', () => {
        // console.log(wrapper.state())
        expect(wrapper.state())
    })

    test('forceUpdateHandler call reset function', ()=>{
        const mock = wrapper.instance().forceUpdateHandler();
        // console.log(mock)
        expect(mock).toEqual(undefined)
    })

    test('set start position', () => {
        const mock = wrapper.instance().setStartPosition();
        // console.log(mock) // if negative, the sprite will go up; if zero, the sprite will be in its bottom, showing the last picture
        expect(mock).toBeLessThanOrEqual(0);
    })

    test('Is the background moving correctly', () => {
        const newPosition = wrapper.state().position - wrapper.instance().speed;
        // console.log(newPosition); // the new position should be greater, because the sprite is always going up
        expect(wrapper.state().position).toBeGreaterThanOrEqual(newPosition);
    })

    test('getting symbol from position', () => {
        const mock = wrapper.instance().onFinish;
        // console.log(mock) // the symbol doesn't stopped yet, only in MainPage
        expect(mock).toBeUndefined();
    })

})

// it('should return the initial state', () => {
    // console.log(wrapper.state({lastPosition, }))
    // console.log(wrapper.state().timeRemaining)
    // console.log(wrapper.instance().moveBackground())
    // console.log(wrapper.instance().forceUpdateHandler)
    // expect(wrapper.instance().forceUpdateHandler()).toBeUndefined()
    // console.log(wrapper.instance().reset())
    // console.log(wrapper.instance().tick())
    // console.log(wrapper.instance().timer)
    // expect(wrapper.instance().timer).toEqual(5)
    // console.log(wrapper.instance().getSymbolFromPosition())
    // console.log(wrapper.instance().currentPosition)
// })
