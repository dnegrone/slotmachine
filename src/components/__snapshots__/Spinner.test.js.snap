// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner Tests renders MainPage without crashing 1`] = `"<div style={{...}} className=\\"icons\\" />"`;
